File: internal/server/threads.go
package server

// A question can be asked without disturbing the conversation it came out of:
// it opens a side thread, which starts as a copy of that conversation and
// goes its own way. The tree of threads lives here — who came from whom, what
// each one is about, which ones are finished, and what a finished one left
// for the thread above it to hear about on its next question.

import (
	"errors"
	"fmt"
	"strings"
	"sync"

	"github.com/mhrlife/nutshell/internal/agent"
)

// threadTitle caps the subject kept for a side thread.
const threadTitle = 120

// errNoThread is what an id nobody opened comes back as.
var errNoThread = errors.New("no such thread")

// errThreadDone is what a question for a finished thread comes back as.
var errThreadDone = errors.New("that side thread is finished")

// thread is one conversation in the tree.
type thread struct {
	id     string
	parent string   // "" for the root thread
	title  string   // what this side thread is about, shown in the trail
	about  string   // the passage it was opened from, when it was opened from one
	asked  []string // the questions the user put to it, in the order they asked them
	closed bool
	notes  []agent.Note // conclusions waiting for this thread's next question
}

// threads is the tree, and the only place a thread's state is kept.
type threads struct {
	mu   sync.Mutex
	byID map[string]*thread
	next int
}

func newThreads() *threads {
	root := &thread{id: agent.RootThread}

	return &threads{byID: map[string]*thread{root.id: root}}
}

// fork opens a side thread of parent and returns its id. about is the passage
// it was opened from, if any, and question is what is about to be asked in
// it; between them they say what the thread is about. The conclusions parent
// has not passed on yet are copied into it: the new thread starts as a copy of
// parent's conversation, which has not heard them either.
func (t *threads) fork(parent, about, question string) (string, error) {
	t.mu.Lock()
	defer t.mu.Unlock()

	above, err := t.openLocked(parent)
	if err != nil {
		return "", err
	}

	about = trimTitle(about)

	title := about
	if title == "" {
		title = trimTitle(question)
	}

	t.next++
	child := &thread{
		id:     fmt.Sprintf("t%d", t.next),
		parent: above.id,
		title:  title,
		about:  about,
		notes:  append([]agent.Note(nil), above.notes...),
	}
	t.byID[child.id] = child

	return child.id, nil
}

// record keeps a question the user asked, so that when the thread is finished
// the one above is told what was being asked, not only what came of it.
func (t *threads) record(id, question string) {
	t.mu.Lock()
	defer t.mu.Unlock()

	if th := t.byID[id]; th != nil {
		th.asked = append(th.asked, question)
	}
}

// conclude is the note a thread leaves behind, once conclusion is known.
func (t *threads) conclude(id, conclusion string) agent.Note {
	t.mu.Lock()
	defer t.mu.Unlock()

	th := t.byID[id]
	if th == nil {
		return agent.Note{Conclusion: conclusion}
	}

	return agent.Note{
		Title:      th.about,
		Asked:      append([]string(nil), th.asked...),
		Conclusion: conclusion,
	}
}

// request describes a thread the way the agent needs it, and hands over the
// conclusions waiting on it: they travel with this one question, and a
// question that never arrives is a conclusion nobody needed to hear.
func (t *threads) request(id string) (agent.Thread, []agent.Note, error) {
	t.mu.Lock()
	defer t.mu.Unlock()

	th, err := t.openLocked(id)
	if err != nil {
		return agent.Thread{}, nil, err
	}

	notes := th.notes
	th.notes = nil

	return agent.Thread{ID: th.id, Parent: th.parent}, notes, nil
}

// finish closes a side thread, along with any side thread opened from it that
// is still going: a thread the user is done with is done with whatever hangs
// off it. It returns the thread itself first, then those, so the caller can
// tell each one's parent. The root thread cannot be finished.
func (t *threads) finish(id string) ([]thread, error) {
	t.mu.Lock()
	defer t.mu.Unlock()

	th, err := t.openLocked(id)
	if err != nil {
		return nil, err
	}

	if th.parent == "" {
		return nil, errors.New("the main thread cannot be finished")
	}

	th.closed = true
	closed := []thread{*th}

	// The tree is small and only ever grows, so a sweep per closing costs
	// nothing and saves keeping a list of children in step.
	for again := true; again; {
		again = false

		for _, other := range t.byID {
			if other.closed || other.parent == "" {
				continue
			}

			if parent := t.byID[other.parent]; parent != nil && parent.closed {
				other.closed = true
				closed = append(closed, *other)
				again = true
			}
		}
	}

	return closed, nil
}

// note leaves a finished side thread's conclusion for the thread above it.
func (t *threads) note(id string, n agent.Note) error {
	t.mu.Lock()
	defer t.mu.Unlock()

	th := t.byID[id]
	if th == nil {
		return errNoThread
	}

	th.notes = append(th.notes, n)

	return nil
}

// title is what a thread is about, for logs and for what the agent is told.
func (t *threads) title(id string) string {
	t.mu.Lock()
	defer t.mu.Unlock()

	if th := t.byID[id]; th != nil {
		return th.title
	}

	return ""
}

// openLocked finds a thread that can still be talked to. The caller holds t.mu.
func (t *threads) openLocked(id string) (*thread, error) {
	if id == "" {
		id = agent.RootThread
	}

	th := t.byID[id]
	if th == nil {
		return nil, errNoThread
	}

	if th.closed {
		return nil, errThreadDone
	}

	return th, nil
}

// trimTitle shortens a subject to one line of a sensible length.
func trimTitle(title string) string {
	title = strings.Join(strings.Fields(title), " ")

	if runes := []rune(title); len(runes) > threadTitle {
		return strings.TrimSpace(string(runes[:threadTitle])) + "…"
	}

	return title
}

File: internal/web/static/threads.js
// A question that comes out of an answer — what is that term, why that
// library — can be asked without dragging the conversation through it: it
// opens a side thread, which knows everything said so far and keeps what
// follows to itself. Ask, get somewhere, say you are done, and you are back
// where you were, with the conclusion brought along if you wanted it.
//
// The tree that makes is the server's business. Here there is only ever one
// thread on screen, a trail at the top saying how deep you are, and a line in
// the thread above marking where a side thread was opened.

const ROOT = 'root';

const transcript = document.getElementById('transcript');
const emptyHint = document.getElementById('empty');
const trail = document.getElementById('trail');
const threadBar = document.getElementById('thread-bar');

// Every thread the page knows about, by id. The marker is the line drawn in
// the thread above, which is also the way back into this one.
const threadList = new Map([[ROOT, { id: ROOT, parent: null, title: '', closed: false, marker: null }]]);

let current = ROOT; // the thread on screen
let forkNext = false; // the next question opens a side thread instead of being asked here
let lastSeen = ROOT; // where the conversation was when the backlog ended

function currentThread() { return threadList.get(current); }

function inSideThread() { return current !== ROOT; }

// inClosedThread is true while a finished thread is being read. Everything it
// holds stays on screen; nothing more can be asked of it.
function inClosedThread() { return !!currentThread().closed; }

// pathTo is the trail from the main thread down to id.
function pathTo(id) {
  const path = [];
  for (let node = threadList.get(id); node; node = threadList.get(node.parent)) path.unshift(node);
  return path;
}

// ---- what the stream says --------------------------------------------------

// openThread draws the line marking where a side thread was opened. It
// belongs to the thread above, which is the one the user was in at the time.
function openThread(parent, data) {
  if (threadList.has(data.id)) return;
  const node = document.getElementById('thread-template').content.firstElementChild.cloneNode(true);
  const thread = { id: data.id, parent, title: data.title || '', closed: false, marker: node };
  threadList.set(thread.id, thread);
  node.dataset.thread = parent;
  node.dataset.id = thread.id;
  renderMarker(thread);
  transcript.appendChild(node);
  applyThreads();
  if (synced) enterThread(thread.id); // this page, or another tab, just opened it
  else lastSeen = thread.id;
}

// finishThread records that a side thread is over, and what came of it.
function finishThread(parent, data) {
  const thread = threadList.get(data.id);
  if (!thread) { logIssue('warn', 'thread', `a thread this page never saw finished (${data.id})`); return; }
  thread.closed = true;
  thread.conclusion = data.conclusion || '';
  renderMarker(thread);
  if (synced) { if (current === thread.id) enterThread(parent); }
  else if (lastSeen === thread.id) lastSeen = parent;
}

// noteThread follows where the conversation went while the backlog replays,
// so a page that reloads comes back where it left off.
function noteThread(id) { if (!synced) lastSeen = id || ROOT; }

// resumeThread is called once the backlog is on screen.
function resumeThread() {
  const thread = threadList.get(lastSeen);
  enterThread(thread && !thread.closed ? lastSeen : (thread ? thread.parent : ROOT) || ROOT);
}

// ---- moving between threads ------------------------------------------------

function enterThread(id) {
  if (!threadList.has(id)) return;
  current = id;
  stopPlayback();
  applyThreads();
  renderTrail();
  renderThreadBar();
  selectLastTurn();
  scrollToEnd();
  // The footer belongs to the thread you are in — a finished one takes no
  // more questions, the one above it does — so it is redrawn on every move.
  renderState();
}

// applyThreads hides everything that belongs to another thread. The
// transcript holds every turn of every thread in the order they happened, so
// showing one thread is a matter of what is on screen, never of what is kept.
function applyThreads() {
  for (const node of transcript.querySelectorAll('[data-thread]')) {
    node.hidden = node.dataset.thread !== current;
  }
  emptyHint.hidden = !!transcript.querySelector('[data-thread]:not([hidden])');
}

// selectLastTurn puts the newest answer of this thread in the document pane.
function selectLastTurn() {
  const here = turns.filter((x) => x.thread === current);
  const answered = here.filter((x) => x.answer);
  select(answered[answered.length - 1] || here[here.length - 1] || null);
}

// ---- opening and finishing -------------------------------------------------

// setFork arms (or disarms) the next question: with it on, the question opens
// a side thread instead of being asked here.
function setFork(on) {
  forkNext = !!on && !currentThread().closed;
  renderForkButton();
  renderState();
}

function toggleFork() { setFork(!forkNext); }

// forking is what ask() sends, and it is spent on the question that uses it.
function forking() { return forkNext; }

function spendFork() { setFork(false); }

// finishHere closes the thread on screen. With inject the agent is asked what
// the thread settled first, and that answer is the only thing the thread
// above ever hears about it.
async function finishHere(inject) {
  if (!inSideThread()) return;
  const thread = currentThread();
  if (thread.closed) { enterThread(thread.parent); return; }
  if (inject && state === 'working') { warn('close', 'stillWorking'); return; }
  setFork(false);
  try {
    await postJSON('/api/close', { thread: thread.id, inject: !!inject, lang: settings.lang });
  } catch (err) {
    fail('close', 'closeFailed', err);
    return;
  }
  if (!inject) return; // closing without a conclusion is done the moment it is recorded
  renderThreadBar(); // the closing turn is running; the way up opens when it lands
}

// ---- drawing ---------------------------------------------------------------

function renderMarker(thread) {
  const node = thread.marker;
  if (!node) return;
  node.classList.toggle('closed', thread.closed);
  node.querySelector('.side-label').textContent = thread.closed ? t('threadClosed') : t('threadOpen');
  const title = node.querySelector('.side-title');
  title.textContent = oneLine(thread.title) || t('threadUntitled');
  title.dir = isRTL(thread.title) ? 'rtl' : 'ltr';
  const note = node.querySelector('.side-note');
  note.textContent = thread.conclusion || '';
  note.dir = isRTL(thread.conclusion) ? 'rtl' : 'ltr';
  note.hidden = !thread.conclusion;
}

// crumbLabel keeps the trail readable at a glance: a few words of what the
// thread is about, never the whole question it started with.
function crumbLabel(thread) {
  if (thread.id === ROOT) return t('threadMain');
  const title = oneLine(thread.title);
  if (!title) return t('threadUntitled');
  return title.length > 36 ? `${title.slice(0, 36)}…` : title;
}

function renderTrail() {
  const path = pathTo(current);
  trail.hidden = path.length < 2;
  if (trail.hidden) return;
  trail.innerHTML = '';
  path.forEach((thread, i) => {
    if (i) trail.appendChild(document.createElement('i')).className = 'trail-sep';
    const crumb = document.createElement('button');
    crumb.type = 'button';
    crumb.className = 'crumb';
    crumb.dataset.id = thread.id;
    crumb.textContent = crumbLabel(thread);
    crumb.dir = thread.id === ROOT ? (document.documentElement.dir || 'ltr') : (isRTL(thread.title) ? 'rtl' : 'ltr');
    crumb.disabled = i === path.length - 1;
    trail.appendChild(crumb);
  });
}

function renderThreadBar() {
  const thread = currentThread();
  threadBar.hidden = !inSideThread();
  if (threadBar.hidden) return;
  const closed = thread.closed;
  threadBar.querySelector('.thread-note').textContent = closed ? t('threadIsClosed') : t('threadHere');
  const done = threadBar.querySelector('.thread-done');
  const drop = threadBar.querySelector('.thread-drop');
  done.textContent = closed ? t('threadBack') : t('threadDone');
  done.title = closed ? '' : t('threadDoneNote');
  drop.hidden = closed;
  drop.textContent = t('threadDrop');
  drop.title = t('threadDropNote');
}

function renderForkButton() {
  el.fork.innerHTML = ICONS.branch;
  el.fork.classList.toggle('on', forkNext);
  el.fork.title = forkNext ? t('forkOnNote') : t('forkNote');
}

// renderThreads redraws everything here, and is what the language switch calls.
function renderThreads() {
  for (const thread of threadList.values()) renderMarker(thread);
  renderTrail();
  renderThreadBar();
  renderForkButton();
}

// ---- wiring ----------------------------------------------------------------

trail.addEventListener('click', (e) => {
  const crumb = e.target.closest('.crumb');
  if (crumb) enterThread(crumb.dataset.id);
});

threadBar.addEventListener('click', (e) => {
  if (e.target.closest('.thread-done')) finishHere(true);
  else if (e.target.closest('.thread-drop')) finishHere(false);
});

document.getElementById('fork').addEventListener('click', toggleFork);

transcript.addEventListener('click', (e) => {
  const marker = e.target.closest('.side');
  if (marker) enterThread(marker.dataset.id);
});
